From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ken Russell <kbr@chromium.org>
Date: Thu, 10 Sep 2026 18:28:29 -0700
Subject: GL: apply host-twiddled TexSubImage2D workaround to all formats.

The previous use_tex_sub_image_for_host_twiddled_npot_uploads
workaround is needed for more texture formats, so simply apply it to
all formats on affected GPUs, and rename the workaround to
use_tex_sub_image_for_client_data_npot_uploads .

Bug: chromium:559320837
Change-Id: I588a6b0c664447e591bc2edb7a7682a103375ae2
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/8392118
Auto-Submit: Kenneth Russell <kbr@chromium.org>
Commit-Queue: Kenneth Russell <kbr@chromium.org>
Reviewed-by: Zhenyao Mo <zmo@chromium.org>
Reviewed-by: Geoff Lang <geofflang@chromium.org>

Electron adaptation notes (44-x-y):
Test-only: the Texture2DTestES3_NPOTHostTwiddledTexture ->
Texture2DTestES3_NPOTClientDataTexture rename in TextureTest.cpp was
applied by hand inside a conflict region shared with the
Texture2DTestES3_OversizedMipLevels suite this branch already carries.

diff --git a/include/platform/autogen/FeaturesGL_autogen.h b/include/platform/autogen/FeaturesGL_autogen.h
index 03e028fa33679affa9a9ede5c38d5c9f974412a4..9721bfe9d9df9afca7931c4fb0d3c1af474072d0 100644
--- a/include/platform/autogen/FeaturesGL_autogen.h
+++ b/include/platform/autogen/FeaturesGL_autogen.h
@@ -788,8 +788,8 @@ struct FeaturesGL : FeatureSetBase
         &members,
     };
 
-    FeatureInfo useTexSubImageForHostTwiddledNpotUploads = {
-        "useTexSubImageForHostTwiddledNpotUploads",
+    FeatureInfo useTexSubImageForClientDataNpotUploads = {
+        "useTexSubImageForClientDataNpotUploads",
         FeatureCategory::OpenGLWorkarounds,
         &members,
     };
diff --git a/include/platform/gl_features.json b/include/platform/gl_features.json
index fafb599da133f79b209d96b1525673dd15efc612..8e3a635477a6df701ca81e2e1c4499ff4ce3ff15 100644
--- a/include/platform/gl_features.json
+++ b/include/platform/gl_features.json
@@ -1039,10 +1039,10 @@
             "issue": "http://crbug.com/546252753"
         },
         {
-            "name": "use_tex_sub_image_for_host_twiddled_npot_uploads",
+            "name": "use_tex_sub_image_for_client_data_npot_uploads",
             "category": "Workarounds",
             "description": [
-                "For texture formats which would be twiddled on the host CPU, upload via TexSubImage2D instead of TexImage2D."
+                "For non-power-of-two textures, upload via TexSubImage2D instead of TexImage2D when uploading client-side data."
             ],
             "issue": "http://crbug.com/553172761"
         }
diff --git a/src/libANGLE/renderer/gl/TextureGL.cpp b/src/libANGLE/renderer/gl/TextureGL.cpp
index 767bcb3358a8bd6b432d0b4303cf04771af6fd9d..48d1f63632066fee1e262a6ceae0c6c18fa888d4 100644
--- a/src/libANGLE/renderer/gl/TextureGL.cpp
+++ b/src/libANGLE/renderer/gl/TextureGL.cpp
@@ -113,13 +113,6 @@ LevelInfoGL GetLevelInfo(const angle::FeaturesGL &features,
                        GetEmulatedAlphaChannel(features, originalInternalFormat));
 }
 
-bool IsHostTwiddledFormat(GLenum internalFormat, GLenum format, GLenum type)
-{
-    return (internalFormat == GL_RGB10_A2 && format == GL_RGBA &&
-            type == GL_UNSIGNED_INT_2_10_10_10_REV) ||
-           (internalFormat == GL_SRGB8_ALPHA8 && format == GL_RGBA && type == GL_UNSIGNED_BYTE);
-}
-
 gl::Texture::DirtyBits GetLevelWorkaroundDirtyBits()
 {
     gl::Texture::DirtyBits bits;
@@ -360,10 +353,8 @@ angle::Result TextureGL::setImageHelper(const gl::Context *context,
     if (nativegl::UseTexImage2D(getType()))
     {
         ASSERT(size.depth == 1);
-        if (features.useTexSubImageForHostTwiddledNpotUploads.enabled && pixels != nullptr &&
-            (!gl::isPow2(size.width) || !gl::isPow2(size.height)) &&
-            IsHostTwiddledFormat(texImageFormat.internalFormat, texImageFormat.format,
-                                 texImageFormat.type))
+        if (features.useTexSubImageForClientDataNpotUploads.enabled && pixels != nullptr &&
+            (!gl::isPow2(size.width) || !gl::isPow2(size.height)))
         {
             ANGLE_GL_TRY_ALWAYS_CHECK(
                 context, functions->texImage2D(
diff --git a/src/libANGLE/renderer/gl/renderergl_utils.cpp b/src/libANGLE/renderer/gl/renderergl_utils.cpp
index ff95994a5c9c817d34d546360302d4938179f978..a9ff1f7068f4aa9802c7a43eb847d805ddda52f1 100644
--- a/src/libANGLE/renderer/gl/renderergl_utils.cpp
+++ b/src/libANGLE/renderer/gl/renderergl_utils.cpp
@@ -2812,7 +2812,7 @@ void InitializeFeatures(const FunctionsGL *functions, angle::FeaturesGL *feature
                             IsPowerVR(vendor) && IsAndroid());
 
     // crbug.com/553172761
-    ANGLE_FEATURE_CONDITION(features, useTexSubImageForHostTwiddledNpotUploads,
+    ANGLE_FEATURE_CONDITION(features, useTexSubImageForClientDataNpotUploads,
                             false /* IsPowerVR(vendor) */);
 
     // Mac Intel drivers are unable to allocate buffers larger than ~1gb
diff --git a/src/tests/gl_tests/TextureTest.cpp b/src/tests/gl_tests/TextureTest.cpp
index 9cb3b88d6c2d9b50e699bdcf168c4b9735555415..293c94dc56819121bec3b6c34145d6db4bf2aa01 100644
--- a/src/tests/gl_tests/TextureTest.cpp
+++ b/src/tests/gl_tests/TextureTest.cpp
@@ -24207,12 +24207,12 @@ TEST_P(Texture2DTestES3_ResetTexStorage2DBaseLevel, TexStorage2DWithNonZeroBaseL
     }
 }
 
-class Texture2DTestES3_NPOTHostTwiddledTexture : public Texture2DTestES3
+class Texture2DTestES3_NPOTClientDataTexture : public Texture2DTestES3
 {};
 
 // Test that non-power-of-two uploads of RGB10_A2 with UNSIGNED_INT_2_10_10_10_REV data succeed and
 // verify texture contents.
-TEST_P(Texture2DTestES3_NPOTHostTwiddledTexture, RGB10A2)
+TEST_P(Texture2DTestES3_NPOTClientDataTexture, RGB10A2)
 {
     constexpr GLsizei kWidth  = 65;
     constexpr GLsizei kHeight = 64;
@@ -24278,7 +24278,7 @@ TEST_P(Texture2DTestES3_NPOTHostTwiddledTexture, RGB10A2)
 
 // Test that non-power-of-two uploads of SRGB8_ALPHA8 with UNSIGNED_BYTE data succeed and
 // verify texture contents, including non-default unpack alignment and skip pixels.
-TEST_P(Texture2DTestES3_NPOTHostTwiddledTexture, SRGB8Alpha8)
+TEST_P(Texture2DTestES3_NPOTClientDataTexture, SRGB8Alpha8)
 {
     constexpr GLsizei kWidth    = 255;
     constexpr GLsizei kHeight   = 256;
@@ -24419,11 +24419,11 @@ ANGLE_INSTANTIATE_TEST_ES3_AND(Texture2DTestES3_ResetTexStorage2DBaseLevel,
                                ES3_OPENGLES().enable(Feature::ResetTexStorage2DBaseLevel),
                                ES3_OPENGLES().disable(Feature::ResetTexStorage2DBaseLevel));
 
-GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(Texture2DTestES3_NPOTHostTwiddledTexture);
+GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(Texture2DTestES3_NPOTClientDataTexture);
 ANGLE_INSTANTIATE_TEST_ES3_AND(
-    Texture2DTestES3_NPOTHostTwiddledTexture,
-    ES3_OPENGL().enable(Feature::UseTexSubImageForHostTwiddledNpotUploads),
-    ES3_OPENGLES().enable(Feature::UseTexSubImageForHostTwiddledNpotUploads));
+    Texture2DTestES3_NPOTClientDataTexture,
+    ES3_OPENGL().enable(Feature::UseTexSubImageForClientDataNpotUploads),
+    ES3_OPENGLES().enable(Feature::UseTexSubImageForClientDataNpotUploads));
 
 GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(TextureSizeLimitTest);
 ANGLE_INSTANTIATE_TEST(TextureSizeLimitTest,
diff --git a/util/autogen/angle_features_autogen.cpp b/util/autogen/angle_features_autogen.cpp
index d5979960159ff52c6b4a0a63945d1d3dd9b8f219..9c682220e55b94d654009cc0ee814cbd1b3a55a7 100644
--- a/util/autogen/angle_features_autogen.cpp
+++ b/util/autogen/angle_features_autogen.cpp
@@ -522,7 +522,7 @@ constexpr PackedEnumMap<Feature, const char *> kFeatureNames = {{
     {Feature::UseStencilTestEnableDynamicState, "useStencilTestEnableDynamicState"},
     {Feature::UseSystemMemoryForConstantBuffers, "useSystemMemoryForConstantBuffers"},
     {Feature::UseTempForNonZeroBaseLevelGenMipmapUsingCopyImageSubData, "useTempForNonZeroBaseLevelGenMipmapUsingCopyImageSubData"},
-    {Feature::UseTexSubImageForHostTwiddledNpotUploads, "useTexSubImageForHostTwiddledNpotUploads"},
+    {Feature::UseTexSubImageForClientDataNpotUploads, "useTexSubImageForClientDataNpotUploads"},
     {Feature::UseUnusedBlocksWithStandardOrSharedLayout, "useUnusedBlocksWithStandardOrSharedLayout"},
     {Feature::UseVertexInputBindingStrideDynamicState, "useVertexInputBindingStrideDynamicState"},
     {Feature::UseVkEventForBufferBarrier, "useVkEventForBufferBarrier"},
diff --git a/util/autogen/angle_features_autogen.h b/util/autogen/angle_features_autogen.h
index ae6bf1311139f775a117536b35e9120ce615e25b..64f595c92bbffecdccfbcd10c341361ca587e48e 100644
--- a/util/autogen/angle_features_autogen.h
+++ b/util/autogen/angle_features_autogen.h
@@ -522,7 +522,7 @@ enum class Feature
     UseStencilTestEnableDynamicState,
     UseSystemMemoryForConstantBuffers,
     UseTempForNonZeroBaseLevelGenMipmapUsingCopyImageSubData,
-    UseTexSubImageForHostTwiddledNpotUploads,
+    UseTexSubImageForClientDataNpotUploads,
     UseUnusedBlocksWithStandardOrSharedLayout,
     UseVertexInputBindingStrideDynamicState,
     UseVkEventForBufferBarrier,
